File: go/luhn/luhn.go
package luhn

import (
	"strings"
	"unicode"
)

func Valid(id string) bool {

	var sum int
	j := 1

	id = strings.ReplaceAll(id, " ", "")

	if len(id) <= 1 {
		return false
	}

	for i := len(id) - 1; i >= 0; i-- {
		if !unicode.IsNumber(rune(id[i])) {
			return false
		}
		digit := int(id[i] - '0')
		if j%2 == 0 {
			if (digit * 2) > 9 {
				sum += (digit * 2) - 9
			} else {
				sum += (digit * 2)
			}
			j++
		} else {
			sum += digit
			j++
		}
	}

	return sum%10 == 0
}

File: typescript/luhn/luhn.ts
export function valid(digitString: string): boolean {
    if (
        (digitString.match(/\d/g) || []).length <= 1 ||
        digitString.match(/[^\d\s]/)
    ) {
        return false;
    }

    digitString = digitString.replaceAll(" ", "");
    let digitArray = Array.from(digitString, (x) => Number.parseInt(x));

    let total = 0;
    const startIndex = digitArray.length % 2 === 0 ? 0 : 1;

    total = digitArray.reduce((sum, digit, index) => {
        if (index % 2 === startIndex) {
            digit = digit * 2 > 9 ? digit * 2 - 9 : digit * 2;
        }

        return sum + digit;
    }, 0);

    return total % 10 === 0;
}
